### PR TITLE
[ILI9XXX] Improve display transfer performance & add second FB

### PR DESCRIPTION
**Pull request by Fabian-Schmidt**
# What does this implement/fix?

Improve transfer speed to SPI display - example WT32-SC01 480x320px display 16 bit color from ~215ms to 105ms.

Use improvement to add a second framebuffer so full display redraw worst case takes ~170ms. Shortcut for full screen refresh is ~115ms.

Second framebuffer is optional and used to detect changes by comparing last display to current display. Goal is to make writing of display lambda functions easier. As I can now use `fill` operation etc.

Typical ESP32 devices with display have PSRam. Smallest possible PSRam is 2MB. One framebuffer with 480x320px and 16 bit color uses 300kb. So there should always be enough memory for second framebuffer.

All performance gain only on ESP32 Arduino. ESP32 IDF SPI is very very slow.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ili9xxx/ili9xxx_display.cpp [hunk 1]
```diff
@@ -185,15 +262,16 @@ void ILI9XXXDisplay::display_() {
...
   for (uint16_t row = 0; row < h; row++) {
     uint32_t pos = start_pos + (row * width_);
+    uint32_t pos_16 = pos + pos;
     uint32_t rem = w;
 
     while (rem > 0) {
       uint32_t sz = std::min(rem, ILI9XXX_TRANSFER_BUFFER_SIZE);
       // ESP_LOGVV(TAG, "Send to display(pos:%d, rem:%d, zs:%d)", pos, rem, sz);
-      buffer_to_transfer_(pos, sz);
+      buffer_to_transfer_(pos, pos_16, sz);
```
Comment (nielsnl68) [inline esphome/components/ili9xxx/ili9xxx_display.cpp]: `
      this->buffer_to_transfer_(pos, pos_16, sz);
`

to make it more consistent met the other changes you made below.

File: esphome/components/ili9xxx/ili9xxx_display.cpp [hunk 2]
```diff
@@ -208,45 +286,70 @@ void ILI9XXXDisplay::display_() {
...
   this->x_high_ = 0;
   this->y_high_ = 0;
 }
 
-uint32_t ILI9XXXDisplay::buffer_to_transfer_(uint32_t pos, uint32_t sz) {
-  for (uint32_t i = 0; i < sz; ++i) {
-    switch (this->buffer_color_mode_) {
-      case BITS_8_INDEXED:
-        transfer_buffer_[i] = display::ColorUtil::color_to_565(
+uint32_t ILI9XXXDisplay::buffer_to_transfer_(uint32_t pos, uint32_t pos_16, uint32_t sz) {
```
Comment (nielsnl68) [inline esphome/components/ili9xxx/ili9xxx_display.cpp]: Not sure of i see the  benefits to send `pos_16` as parameter in compare with just create the variable only when you are in `BITS_16` mode.

**Review by nielsnl68 — CHANGES_REQUESTED** [review]
Thanks @Fabian-Schmidt for taking over the touch to make the ili9xxx platform component much faster and better.

I have added 2 small change requests.

File: esphome/components/ili9xxx/ili9xxx_display.cpp (hunk 1 above)
Comment (Fabian-Schmidt) [reply esphome/components/ili9xxx/ili9xxx_display.cpp]: fixed

File: esphome/components/ili9xxx/ili9xxx_display.cpp (hunk 2 above)
Comment (Fabian-Schmidt) [reply esphome/components/ili9xxx/ili9xxx_display.cpp]: I was playing with performance optimizations and this one is not worth it. Removed the parameter.

**Comment by Fabian-Schmidt** [comment]
Thanks @nielsnl68 for this very quick feedback on the changes.

**Pull request closed without merging**